test/hotspot/jtreg/serviceability/jvmti/SuspendWithObjectMonitorWait/SuspendWithObjectMonitorWait.java

```diff
@@ -265,18 +293,266 @@ public int doWork(int timeMax, PrintStream out) {
+                } catch(Exception e) {}
+            }
+
+            try {
+                resumer.join(JOIN_MAX * 1000);
+                if (resumer.isAlive()) {
+                    System.err.println("Failure at " + count + " loops.");
+                    throw new InternalError("resumer thread is stuck");
+                }
+                waiter.join(JOIN_MAX * 1000);
+                if (waiter.isAlive()) {
+                    System.err.println("Failure at " + count + " loops.");
+                    throw new InternalError("waiter thread is stuck");
+                }
+            } catch (InterruptedException ex) {
+            }
+
+            checkTestState(TS_WAITER_DONE);
+        }
+
+        System.out.println("Executed " + count + " loops in " + timeMax +
+                " seconds.");
+
+        return 0;
+    }
+
+    // Suspend on the re-entry path of wait.
+    public int doWork3(int timeMax, PrintStream out) {
+        SuspendWithObjectMonitorWaitWorker waiter;    // waiter thread
+        SuspendWithObjectMonitorWaitWorker resumer;    // resumer thread
+
+        System.out.println("About to execute for " + timeMax + " seconds.");
+
+        long start_time = System.currentTimeMillis();
+        while (System.currentTimeMillis() < start_time + (timeMax * 1000)) {
+            count++;
+            testState = TS_INIT;  // starting the test loop
+
+            // launch the waiter thread
+            synchronized (barrierLaunch) {
+                waiter = new SuspendWithObjectMonitorWaitWorker("waiter", 1);
+                waiter.start();
+
+                while (testState != TS_WAITER_RUNNING) {
+                    try {
+                        barrierLaunch.wait(0);  // wait until it is running
+                    } catch (InterruptedException ex) {
+                    }
+                }
+            }
+
+            // launch the resumer thread
+            synchronized (barrierLaunch) {
+                resumer = new SuspendWithObjectMonitorWaitWorker("resumer", waiter);
+                resumer.start();
+
+                while (testState != TS_RESUMER_RUNNING) {
+                    try {
+                        barrierLaunch.wait(0);  // wait until it is running
+                    } catch (InterruptedException ex) {
+                    }
+                }
+            }
+
+            try {
+                Thread.sleep(1000);
+            } catch(Exception e) {}
+
+            checkTestState(TS_RESUMER_RUNNING);
+
+            // The waiter thread was synchronized on threadLock before it
+            // set TS_WAITER_RUNNING and notified barrierLaunch above so
+            // we cannot enter threadLock until the waiter thread calls
+            // threadLock.wait().
+            synchronized (threadLock) {
+                // notify waiter thread so it can try to reenter threadLock
+                testState = TS_READY_TO_NOTIFY;
+                threadLock.notify();
+
```

Comment:
We should add a `Thread.sleep` here to give time for the wait to timeout before suspending the waiter thread. Should be more than what the waiter thread is using as timeout. I have tested a timeout value of 100ms and here sleeping 200ms and always see the deadlock on `reenter_internal` (with current mainline code).